src/query/v2/multiframe.cpp

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "query/v2/multiframe.hpp"
+
+#include "query/v2/bindings/frame.hpp"
+#include "utils/pmr/vector.hpp"
+
+namespace memgraph::query::v2 {
+
+MultiFrame::MultiFrame(FrameWithValidity default_frame, size_t number_of_frames,
+                       utils::MemoryResource *execution_memory)
+    : default_frame_(default_frame),
+      frames_(utils::pmr::vector<FrameWithValidity>(number_of_frames, default_frame, execution_memory)) {
+  MG_ASSERT(number_of_frames > 0);
+  MG_ASSERT(!default_frame.IsValid());
+}
+
+MultiFrame::MultiFrame(const MultiFrame &other) : default_frame_(other.default_frame_) {
+  /*
+  TODO
+  Do we just copy all frames or do we make distinctions between valid and not valid frames? Does it make any
+  difference?
+  */
+  frames_.reserve(other.frames_.size());
+  std::transform(other.frames_.begin(), other.frames_.end(), std::back_inserter(frames_),
+                 [&default_frame = default_frame_](const auto &other_frame) {
+                   if (other_frame.IsValid()) {
+                     return other_frame;
+                   }
+                   return default_frame;
+                 });
+}
+
+// NOLINTNEXTLINE (bugprone-exception-escape)
+MultiFrame::MultiFrame(MultiFrame &&other) noexcept
+    : default_frame_(std::move(other.default_frame_)), frames_(std::move(other.frames_)) {}
+
+void MultiFrame::ResetAllFramesInvalid() noexcept {
+  std::for_each(frames_.begin(), frames_.end(), [](auto &frame) { frame.MakeInvalid(); });
+}
+
+bool MultiFrame::HasValidFrame() const noexcept {
+  return std::any_of(frames_.begin(), frames_.end(), [](auto &frame) { return frame.IsValid(); });
+}
```

Comment:
I wouldn't be so sure these are `noexcept` https://en.cppreference.com/w/cpp/algorithm/for_each#Exceptions

Reply:
I would tend to think it's fine as we are not allocating any memory here and isValid/MakeInvalid do not throw exception
(Ps: i have just marked MakeInvalid and IsValid as noexpect)